ibis/common/bases.py

```diff
@@ -0,0 +1,189 @@
+    provides a `__create__` classmethod that can be used to change the class
+    instantiation behavior. Also enables weak references for the subclasses.
+    """
+
+    __slots__ = ("__weakref__",)
+    __create__ = classmethod(type.__call__)  # type: ignore
+
+
+class Immutable(Base):
+    """Prohibit attribute assignment on the instance."""
+
+    def __copy__(self):
+        return self
+
+    def __deepcopy__(self, memo):
+        return self
+
+    def __setattr__(self, name: str, _: Any) -> None:
+        raise AttributeError(
+            f"Attribute {name!r} cannot be assigned to immutable instance of "
+            f"type {type(self)}"
+        )
+
+
+class Singleton(Base):
+    """Cache instances of the class based on instantiation arguments."""
+
+    __instances__: Mapping[Any, Self] = WeakValueDictionary()
+
+    @classmethod
+    def __create__(cls, *args, **kwargs):
+        key = (cls, args, FrozenDict(kwargs))
+        try:
+            return cls.__instances__[key]
+        except KeyError:
+            instance = super().__create__(*args, **kwargs)
+            cls.__instances__[key] = instance
+            return instance
+
+
+class Final(Base):
+    """Prohibit subclassing."""
+
+    def __init_subclass__(cls, **kwargs):
+        cls.__init_subclass__ = cls.__prohibit_inheritance__
+
+    @classmethod
+    def __prohibit_inheritance__(cls, **kwargs):
+        raise TypeError(f"Cannot inherit from final class {cls}")
+
+
+class Comparable(Base):
+    """Enable quick equality comparisons.
+
+    The subclasses must implement the `__equals__` method that returns a boolean
+    value indicating whether the two instances are equal. This method is called
+    only if the two instances are of the same type and the result is cached for
+    future comparisons.
+
+    Since the class holds a global cache of comparison results, it is important
+    to make sure that the instances are not kept alive longer than necessary.
+    This is done automatically by using weak references for the compared objects.
+    """
+
+    __cache__ = WeakCache()
+
+    def __eq__(self, other) -> bool:
+        try:
+            return self.__cached_equals__(other)
+        except TypeError:
+            return NotImplemented
+
+    @abstractmethod
+    def __equals__(self, other) -> bool:
+        ...
+
+    def __cached_equals__(self, other) -> bool:
+        if self is other:
+            return True
+
+        # type comparison should be cheap
+        if type(self) is not type(other):
+            return False
+
+        # reduce space required for commutative operation
+        if id(self) < id(other):
+            key = (self, other)
+        else:
+            key = (other, self)
+
+        try:
+            result = self.__cache__[key]
+        except KeyError:
+            result = self.__equals__(other)
+            self.__cache__[key] = result
+
+        return result
+
+
+class Slotted:
```

Comment:
Should this subclass from Base?

Reply:
Reading further I guess this intentional? Can you add a line in the docstrings saying this isn't a subclass of `Base`?

Reply:
Yes, definitely. I've done too much rebasing and it got lost somewhere. Updating.

Reply:
Updated.